ls1intum/Artemis
# Fix index out of bounds Exception in ProgrammingSubmissionService

### Checklist

### Motivation and Context


In some case a submission for a programming exercise might not have a result, even no automatic result. If this event comes to pass, which we did not consider before, the `lockSubmission()` method in the `ProgrammingSubmissionService` tries to access the submissions' result with id `-1`. 

**EDIT:** The issue can only happen with older programming exercises. Before the addition of multiple results, the new manual result was in a separate submission. Therefore, the sumbission only had one result when it got loaded from the databse. If the assessment of such submissions is canceled now, it leaves behind a submission without any results, which leads to the mentioned problem.  With newly created Programming Exercises, this issue will not happen, as there will always be an automatic result.(Unless it is deleted manually, and this case is covered now, too.)

<details>
<summary>Error log</summary>
<p>

```c#
Jan 30 11:15:20 vmbruegge105 java[27609]: java.lang.IndexOutOfBoundsException: Index -1 out of bounds for length 0
Jan 30 11:15:20 vmbruegge105 java[27609]:         at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at java.base/java.util.Objects.checkIndex(Objects.java:359)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at java.base/java.util.ArrayList.get(ArrayList.java:427)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at de.tum.in.www1.artemis.domain.Submission.getResultForCorrectionRound(Submission.java:127)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at de.tum.in.www1.artemis.service.ProgrammingSubmissionService.lockSubmission(ProgrammingSubmissionService.java:701)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at de.tum.in.www1.artemis.service.ProgrammingSubmissionService.lockAndGetProgrammingSubmissionWithoutResult(ProgrammingSubmissionService.java:679)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at de.tum.in.www1.artemis.service.ProgrammingSubmissionService$$FastClassBySpringCGLIB$$fbbf5f1c.invoke(<generated>)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at de.tum.in.www1.artemis.service.ProgrammingSubmissionService$$EnhancerBySpringCGLIB$$6fd5313b.lockAndGetProgrammingSubmissionWithoutResult(<generated>)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at de.tum.in.www1.artemis.web.rest.ProgrammingSubmissionResource.getProgrammingSubmissionWithoutAssessment(ProgrammingSubmissionResource.java:435)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at de.tum.in.www1.artemis.web.rest.ProgrammingSubmissionResource$$FastClassBySpringCGLIB$$9996c806.invoke(<generated>)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:779)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:750)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.security.access.intercept.aopalliance.MethodSecurityInterceptor.invoke(MethodSecurityInterceptor.java:61)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:750)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
Jan 30 11:15:20 vmbruegge105 java[27609]:         at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:750)
```

</p>
</details>  

### Description
We added a check to ensure that a submission without any results is handled differently from usual submissions. When assessing it new feedback is created, but it is also clear that no automatic tests ran.
![image](https://user-images.githubusercontent.com/33342534/106392968-98b37580-63f4-11eb-8479-1d629ac9ced4.png)




### Steps for Testing

In the normal case there is either _no submission_ or a _submission that has an automatic result_. (_Unless I really did miss something_) I guess what theoretically could have happened is, that a submission was created, but somehow there was an error somewhere else on bitbucket/bamboo, and no result was received and saved. 
You can either test this in exam-mode or in normal exercise mode, in both I couldn't reproduce the initial problem naturally.
<details>
<summary>Things I have tried to reproduce it naturally:</summary>

- submit the exam but not the exercise
- submit the exercise, but not the exam
- after the exam is over use the "assess unsubmitted exams" feature
- under participations use the "trigger" feature (if the student has used the submit button there is no submission)
- different ways to access the code/assessment editor: "new assessment" from the `exercise-assessment-dashboard` or "next assessment" in the `code-editor`

</details>

What you can do is manipulate the database by hand to get in the situation that causes the error.

1. create an programming exrcise / an exam with an programming exercise
2. participate with a student, submit
3. in MySQL, remove the automatically created result + feedbacks of this submission from your database
4. now you can go to the assessment dashboard -> new assessment
![image](https://user-images.githubusercontent.com/33342534/106394013-8b00ee80-63fa-11eb-84cd-5fbb09392261.png)
5. you will notice that the test part of the resultString is undefined, but the code-editor opens.

On develop the new assessment action leads to the out of bounds exception, and therefore doesn't open the code-editor.
